videostore/mux.go

```diff
@@ -0,0 +1,559 @@
+	nalu, err := h264.AnnexBMarshal(au)
+	if err != nil {
+		m.logger.Errorf("failed to marshal annex b: %s", err)
+		return err
+	}
+	if !m.metadata.firstTimeStampsSet {
+		m.metadata.firstPTS = pts
+		m.metadata.firstDTS = dts
+		m.metadata.firstTimeStampsSet = true
+	}
+	err = m.rawSeg.WritePacket(nalu, pts-m.metadata.firstPTS, dts-m.metadata.firstDTS, isRandomAccess)
+	if err != nil {
+		m.logger.Errorf("error writing packet to segmenter: %s", err)
+	}
+	return nil
+}
+
+func (m *rawSegmenterMux) writeH264(au [][]byte, pts int64) error {
+	var filteredAU [][]byte
+	nonIDRPresent := false
+	idrPresent := false
+
+	for _, nalu := range au {
+		//nolint:mnd
+		typ := h264.NALUType(nalu[0] & 0x1F)
+		switch typ {
+		case h264.NALUTypeSPS:
+			m.metadata.sps = nalu
+			m.metadata.spsUnChanged = false
+			continue
+
+		case h264.NALUTypePPS:
+			m.metadata.pps = nalu
+			continue
+
+		case h264.NALUTypeAccessUnitDelimiter:
+			continue
+
+		case h264.NALUTypeIDR:
+			idrPresent = true
+
+		case h264.NALUTypeNonIDR:
+			nonIDRPresent = true
+		case h264.NALUTypeDataPartitionA,
+			h264.NALUTypeDataPartitionB,
+			h264.NALUTypeDataPartitionC,
+			h264.NALUTypeSEI,
+			h264.NALUTypeEndOfSequence,
+			h264.NALUTypeEndOfStream,
+			h264.NALUTypeFillerData,
+			h264.NALUTypeSPSExtension,
+			h264.NALUTypePrefix,
+			h264.NALUTypeSubsetSPS,
+			h264.NALUTypeReserved16,
+			h264.NALUTypeReserved17,
+			h264.NALUTypeReserved18,
+			h264.NALUTypeSliceLayerWithoutPartitioning,
+			h264.NALUTypeSliceExtension,
+			h264.NALUTypeSliceExtensionDepth,
+			h264.NALUTypeReserved22,
+			h264.NALUTypeReserved23,
+			h264.NALUTypeSTAPA,
+			h264.NALUTypeSTAPB,
+			h264.NALUTypeMTAP16,
+			h264.NALUTypeMTAP24,
+			h264.NALUTypeFUA,
+			h264.NALUTypeFUB:
+		default:
+			return errors.New("invalid nalu")
+		}
+
+		filteredAU = append(filteredAU, nalu)
+	}
+
+	au = filteredAU
+
+	if au == nil || (!nonIDRPresent && !idrPresent) {
+		return nil
+	}
+
+	if err := m.maybeReInitVideoStore(); err != nil {
+		m.logger.Debugf("unable to init video store: %s", err.Error())
+		return nil
+	}
+
+	// add SPS and PPS before access unit that contains an IDR
+	if idrPresent {
+		au = append([][]byte{m.metadata.sps, m.metadata.pps}, au...)
+	}
+
+	if m.metadata.dtsExtractor == nil {
+		// skip samples silently until we find one with a IDR
+		if !idrPresent {
+			return nil
+		}
+		m.metadata.dtsExtractor = h264.NewDTSExtractor2()
+	}
+
+	dts, err := m.metadata.dtsExtractor.Extract(au, pts)
+	if err != nil {
+		m.logger.Debugf("dtsExtractor Extract err: %s", err.Error())
+		return nil
+	}
+
+	packed, err := h264.AnnexBMarshal(au)
+	if err != nil {
+		m.logger.Errorf("AnnexBMarshal err: %s", err.Error())
+		return err
+	}
+
+	if !m.metadata.firstTimeStampsSet {
+		m.metadata.firstPTS = pts
+		m.metadata.firstDTS = dts
+		m.metadata.firstTimeStampsSet = true
+	}
+	err = m.rawSeg.WritePacket(packed, pts-m.metadata.firstPTS, dts-m.metadata.firstDTS, idrPresent)
```

Comment:
Mind adding a comment about why we need to adjust timestamps?

Reply:
yep done